### PR TITLE
HAFNIUM Playbooks

### PR DESCRIPTION
**Pull request by davisshannon**
Playbooks referenced in upcoming blog.

### CODE REVIEW

**Comment by davisshannon** [comment]
Playbooks referenced in upcoming Splunk Security Blog around HAFNIUM Exchange exploits.

File: Terminate_W3WP_Spawned_Processes.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+"""
```
Comment (philroyer-phantom) [inline Terminate_W3WP_Spawned_Processes.py]: Regarding the name of the playbook, I am not seeing anything in here that is specific to W3WP or Hafnium, so I think we can make this something more generic like "terminate_identified_processes" (we also use lowercase playbook names).

Comment (philroyer-phantom) [reply Terminate_W3WP_Spawned_Processes.py]: At the top level, I try to always have a simple description, such as "Use Windows Remote Management to terminate the process for each artifact with a process_id, and terminate any child processes as identified by searching the Endpoint Data Model in Splunk."

File: Terminate_W3WP_Spawned_Processes.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    
+    # call 'Terminate_Process' block
+    Terminate_Process(container=container)
+
+    return
+
+def Format_Splunk_Search(action=None, success=None, container=None, results=None, handle=None, filtered_artifacts=None, filtered_results=None, custom_function=None, **kwargs):
+    phantom.debug('Format_Splunk_Search() called')
+    
+    template = """index=* parent_process_id={0} | fields *"""
```
Comment (philroyer-phantom) [inline Terminate_W3WP_Spawned_Processes.py]: According to the endpoint data model these are numeric identifiers, not GUIDs, so I think we should narrow this search down to just the host in the event.

File: Terminate_W3WP_Spawned_Processes.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    
+    # build parameters list for 'Find_Child_Processes' call
+    parameters.append({
+        'query': formatted_data_1,
+        'command': "search",
+        'display': "",
+        'parse_only': "",
+    })
+
+    phantom.act(action="run query", parameters=parameters, assets=['splunk-1'], callback=Terminate_Child_Processes, name="Find_Child_Processes")
```
Comment (philroyer-phantom) [inline Terminate_W3WP_Spawned_Processes.py]: For uniformity with other playbooks please change this Splunk asset name to "splunk"

File: Terminate_W3WP_Spawned_Processes.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        for results_item_1 in results_data_1:
+            parameters.append({
+                'pid': results_item_1[0],
+                'name': "",
+                'ip_hostname': container_item[0],
+                # context (artifact id) is added to associate results with the artifact
+                'context': {'artifact_id': results_item_1[1]},
+            })
+
+    phantom.act(action="terminate process", parameters=parameters, assets=['windowsrm'], name="Terminate_Child_Processes", parent_action=action)
```
Comment (philroyer-phantom) [inline Terminate_W3WP_Spawned_Processes.py]: For uniformity with other playbooks please change this asset name to "winrm" for this action and the other terminate process action.

File: Delete_Exchange_Webshells.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+"""
```
Comment (philroyer-phantom) [inline Delete_Exchange_Webshells.py]: This does not have any specific actions for Hafnium or webshells so we should give it a more generic name like "delete_suspicious_files".

File: Delete_Exchange_Webshells.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            'command': formatted_data_1,
+            'shell_id': "",
+            'arguments': "",
+            'command_id': "",
+            'ip_hostname': container_item[0],
+            # context (artifact id) is added to associate results with the artifact
+            'context': {'artifact_id': container_item[1]},
+        })
+
+    phantom.act(action="run command", parameters=parameters, assets=['windowsrm'], callback=Format_Del_Command, name="Gather_Shell_Details")
```
Comment (philroyer-phantom) [inline Delete_Exchange_Webshells.py]: Please use "winrm" as the asset name to match existing playbooks.

File: Delete_Exchange_Webshells.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
+"""
+"""
```
Comment (philroyer-phantom) [inline Delete_Exchange_Webshells.py]: We also should have a top level description and tooltips of the key actions to make this easier to read.

**Review by philroyer-phantom — CHANGES_REQUESTED** [review]
The content of the playbooks looks good, but I just had a few adjustments around naming and documentation.

File: Terminate_W3WP_Spawned_Processes.py (hunk 1 above)
Comment (davisshannon) [reply Terminate_W3WP_Spawned_Processes.py]: Just updated.

File: Terminate_W3WP_Spawned_Processes.py (hunk 2 above)
Comment (davisshannon) [reply Terminate_W3WP_Spawned_Processes.py]: Added the host details into the Splunk search.

File: Terminate_W3WP_Spawned_Processes.py (hunk 3 above)
Comment (davisshannon) [reply Terminate_W3WP_Spawned_Processes.py]: Changed the asset to splunk

File: Terminate_W3WP_Spawned_Processes.py (hunk 4 above)
Comment (davisshannon) [reply Terminate_W3WP_Spawned_Processes.py]: changed the asset to winrm

File: Delete_Exchange_Webshells.py (hunk 5 above)
Comment (davisshannon) [reply Delete_Exchange_Webshells.py]: name has been changed

File: Delete_Exchange_Webshells.py (hunk 6 above)
Comment (davisshannon) [reply Delete_Exchange_Webshells.py]: Renamed

File: Delete_Exchange_Webshells.py (hunk 7 above)
Comment (davisshannon) [reply Delete_Exchange_Webshells.py]: Done

**Comment by davisshannon** [comment]
Changes implemented

**Pull request merged**